src/runtime-rs/crates/hypervisor/src/device/driver/port_device.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2024 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+//use std::collections::HashMap;
+
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+
+use crate::device::{
+    hypervisor,
+    topology::{PCIePort, PCIePortBusPrefix, PCIeTopology, TopologyPortDevice},
+    Device, DeviceType,
+};
+
+#[derive(Debug, Default, Clone)]
+pub struct PortDeviceConfig {
+    pub port_type: PCIePort,
+    pub total_ports: u32,
+    pub machine_type: String,
+    pub mem_size_32bit: u64,
+    pub mem_size_64bit: u64,
+}
+
+impl PortDeviceConfig {
+    pub fn new(port_type: PCIePort, total_ports: u32) -> Self {
+        Self {
+            port_type,
+            total_ports,
+            machine_type: "QemuQ35".to_string(),
+            mem_size_32bit: 2097152_u64,
+            mem_size_64bit: 4194304_u64,
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct PCIePortDevice {
+    /// device id for sharefs device in device manager
+    pub device_id: String,
+
+    /// config for sharefs device
+    pub config: PortDeviceConfig,
+}
+
+impl PCIePortDevice {
+    pub fn new(device_id: &str, config: &PortDeviceConfig) -> Self {
+        Self {
+            device_id: device_id.to_string(),
+            config: config.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl Device for PCIePortDevice {
+    async fn attach(
+        &mut self,
+        pcie_topo: &mut Option<&mut PCIeTopology>,
+        h: &dyn hypervisor,
+    ) -> Result<()> {
+        if let Some(topology) = pcie_topo {
+            for index in 0..self.config.total_ports {
+                if let Some(devices) = topology
+                    .pcie_devices_per_port
+                    .get_mut(&self.config.port_type)
+                {
+                    let id = match self.config.port_type {
+                        PCIePort::RootPort => format!("{}{}", PCIePortBusPrefix::RootPort, index),
+                        PCIePort::SwitchPort => {
+                            format!("{}{}", PCIePortBusPrefix::SwitchPort, index)
+                        }
+                        _ => return Err(anyhow!("unspported pcie port type")),
+                    };
+                    devices.push(TopologyPortDevice {
+                        id,
+                        bus: "pcie.0".to_string(),
+                        port: self.config.port_type,
+                        occupied: false,
+                    })
```

Comment:
We're pushing here directly into `PCIeTopology`'s per-port-type device lists (`PCIeTopology::pcie_devices_per_port`) `total_ports` times.  How does that play with `PCIeTopology::pcie_root_ports` and `pcie_switch_ports`, is it guaranteed that we don't make the `PCIeTopology` instance inconsistent?